### PR TITLE
Make sure that loss can be freed after every iteration

### LINKED ISSUE #40714: [Templates] Finetuneing template 04 OOMS due to gpu memory leak when using LoRA + V100s
### What happened + What you expected to happen

[Fine-tuning template 04](https://github.com/ray-project/ray/tree/master/doc/source/templates/04_finetuning_llms_with_deepspeed) OOMs under specific circumstances.

Start an AWS _or_ GCP V100 node.
Deploy the template.
Run `./run_llama_ft.sh --size=7b --lora`.

This will OOM after a while after GRAM fills up. GRAM usage will fill up linearly with steps strongly suggesting a memory leak.
Attempts to reproduce this on AWS's p4des or G5s failed. So this appears to be more or less specific to V100s.

Screenshot shows GRAM (orange) on runs with different context lengths (8, 4, 1 in this order) with LoRA on a V100.
(The final run with batch size 1 grows to approx 100% before crashing).
Screenshot 2023-10-25 at 10 36 51

This is how it looks like on some A100s and should look like (no linear increase, just a flat GRAM curve (orange)).
Screenshot 2023-10-26 at 11 16 26

### Versions / Dependencies

master
https://github.com/ray-project/ray/tree/master/doc/source/templates/04_finetuning_llms_with_deepspeed

### Reproduction script

. 

### Issue Severity

None

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
## Why are these changes needed?

This implements @woshiyyya 's fix to the memory leak in our core training loop.

Closes #40714

### CODE REVIEW

File: doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py [hunk 1]
```diff
@@ -399,7 +399,7 @@ def training_function(kwargs: dict):
                 s_fwd = time.time()
                 outputs = model(**batch)
                 loss = outputs.loss
-                loss_sum += loss
```
Comment (matthewdeng) [inline doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py:402]: @woshiyyya I guess this and `loss.detach()` should end up doing the same thing?

**Review by matthewdeng — APPROVED** [review]
Great job @ArturNiederfahrenhorst @woshiyyya getting to the bottom of this! You guys inspire me. 

<div><br/>via <a href="https://giphy.com/nba/">NBA</a> on <a href="https://giphy.com/gifs/nba-curry-steph-stephcurry-LNnWGuPSnjGZ7CqDfd">GIPHY</a></div>

File: doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py:402]: I was going for loss.detach() and @woshiyyya told me that he used .item() when doing this.
I have since spun up. a workspace to validate @woshiyyya 's findings.

I could run a multiple-epoch training without signs of a leak with the commit here on V100s.
Without, it was OOming.

CC @kouroshHakha

**Comment by kouroshHakha** [comment]
Also kudos to @mak-454 for bringing that line of code to the team's attention.

**Review by kouroshHakha — APPROVED** [review]

File: doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py (hunk 1 above)
Comment (woshiyyya) [reply doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py:402]: Yes they should have the same effect. I saw the GRAM usage are constant during training.
